ENIGMAsystem/SHELL/Platforms/General/PFwindow.cpp

```diff
@@ -85,6 +85,19 @@ void compute_window_size() {
 
 }  //namespace enigma
 
+#ifdef DEBUG_MODE
+  #include "Widget_Systems/widgets_mandatory.h" // for DEBUG_MESSAGE
+  #define KEY_CHECK(key,ret) \
```

Comment:
You joke about using asset array but I am sick of seeing this macro copied everywhere. You could at least write a general set of this macro that can be used like so:

```cpp
#define CHECK_BOUNDS_OR(idx, array, ret) ....
```

Reply:
I invented AssetArray for two reasons, getting rid of the shitty macros was one of them. Considering that, I don't know why you haven't fallen in love with AssetArray yet.

Reply:
I like the idea of asset array but every-time I go to use it I find a bug plus it requiring you implement ~3 methods instead of having a default implementation is a pain in the ass cause you still need to add extra code everywhere

Reply:
I wanted to have a system more like:

```cpp
<template class T>
T get_or(size_t idx, std::vector<T> vec) {
  if (idx > 0 && idx < vec.size()) return vec[idx]
  else  {
  	DEBUG_MESSAGE("SUX2BEU");
  	return T();
  }
}```
and you could add a similar one for std::map or whatever

I think asset array is too over engineered imo.

Reply:
That's the other problem AssetArray tries to solve, cache locality for optimization. It's faster to fetch the next asset if it's right next to it. Maps are R/B trees, and unordered maps are hash tables, neither of which are good for looping all the assets of the same type. We started to move some resources to map as well as reuse ids in various places which was eventually going to be really inconsistent and bite us in the ass, so I jumped on AssetArray so we would have one single concept of how assets are stored in memory. If you still want map or reusing ids or some other behavior, all of that is in a single location. If you think your pattern can be adapted to solve that problem, then it's not out of the question we could use it in place of AssetArray.

Reply:
I don't think it'd be a good idea to replace it at this point but clearly we need some in-between for simpler bounds checks of data structures that don't care about reserving IDs. Either in the form of the function I just gave or a base class asset array expands on.

Reply:
For now, I propose just adding (to AssetArray.h) something like:

```cpp
int get_or_default(int idx, int min, int max, int def, const std::string& debug_msg) {
  if (idx < min || > max) {
   DEBUG_MESSAGE(debug_msg + " value: " + idx);
   return def;
  }
  return idx;
}

#define VALID_OR_RETURNV(val, invalidVal) if (val == invalidVal) return;
```

Then doing something like:

```cpp
const std::string kyb_range_error = "Keyboard constant is outside the normal 0-255 range.";
VALID_OR_RETURNV(get_or_default(i, 0,  vk_last_key, -1), -1);
```